src/Grpc.Net.Client/Balancer/Internal/SocketConnectivitySubchannelTransport.cs

```diff
@@ -337,6 +291,68 @@ private void OnCheckSocketConnection(object? state)
         }
     }
 
+    /// <summary>
+    /// Checks whether the socket is healthy. May read available data into the passed in buffer.
+    /// Returns true if the socket should be closed.
+    /// </summary>
+    private bool CheckSocket(Socket socket, BalancerAddress socketAddress, ref List<ReadOnlyMemory<byte>>? socketData, out Exception? checkException)
+    {
+        checkException = null;
+
+        try
+        {
+            SocketConnectivitySubchannelTransportLog.CheckingSocket(_logger, _subchannel.Id, socketAddress);
+
+            // Poll socket to check if it can be read from. Unfortunatly this requires reading pending data.
+            // The server might send data, such as an HTTP/2 SETTINGS or GOAWAY frame, so we need to read and cache it.
+            //
+            // Available data needs to be read now because the only way to determine whether the connection is closed is to
+            // get the results of polling after available data is received.
+            bool hasReadData;
+            do
+            {
+                if (IsSocketInBadState(socket, socketAddress))
+                {
+                    return true;
+                }
+
+                var available = socket.Available;
+                if (available > 0)
+                {
+                    hasReadData = true;
+                    var serverDataAvailable = CalculateInitialSocketDataLength(socketData) + available;
+                    if (serverDataAvailable > MaximumInitialSocketDataSize)
+                    {
+                        // Data sent to the client before a connection is started shouldn't be large.
+                        // Put a maximum limit on the buffer size to prevent an unexpected scenario from consuming too much memory.
+                        throw new InvalidOperationException($"The server sent {serverDataAvailable} bytes to the client before a connection was established. Maximum allowed data exceeded.");
```

Comment:
Not new but it's a little ugly to throw and then immediately catch this.

Reply:
The exception gets logged and pass out in an out parameter.

Reply:
You could just set the out parameter here, but you'd have to duplicate the logging statement, so it's probably not worth it.